### DISCUSSION TITLE
Transpile to PHP?

### DISCUSSION THREAD

**Discussion by danon** [pfusik/cito#95] [General]
Could fusion be transpiled to PHP?

**Comment by pfusik** [comment]
I have little experience with PHP, but I think adding a PHP backend is feasible. It appears similar to Perl 5 and we had a Perl 5 backend many years ago.

**Reply by danon** [reply]
@pfusik I think I am an expert in PHP (versions 5.3 to 8.2), so if I can be of any help, feel free to reach me. Im the author of this thingy: https://t-regx.com/.

**Comment by pfusik** [comment]
Let's try some manual translations as I'm learning PHP. :)

Consider this Fusion code:

```csharp
class Foo
{
	List<byte>() Buffer;
	internal void Push!(byte b)
	{
		Buffer.Add(b);
	}
	public List<byte> GetBuffer() => Buffer;
}
```

Should this be transpiled to this?

```php
<?php
class Foo
{
	private readonly array $buffer = array();
	function push(int $b): void
	{
		$this->buffer[] = $b;
	}
	function &get_buffer(): array
	{
		return $this->buffer;
	}
}
```

**Reply by danon** [reply]
Okay, so:

1. `readonly array` cannot be modified. Not only reassigned, but modified, so that would essentially be an immutable data structure. if `List<byte>` is mutable, then the property cannot be `readonly`. Frankly, I don't think `readonly` has any particular use for Fusion, other than immutable data structures.'
2. I think, instead of:
   ```php
   private readonly array $buffer = array();
   ```
   we should go with
   ```php
   private array $buffer = [];
   ```
   `[]` is a more standard way to declare arrays now, `array()` was used before PHP 5.4 (which was ages ago).
3. I didn't find any problems with using `&` in method return, provided that its calls will also be used with `&`, so:
   ```
   # fusion
   foo = new Foo();
   buffer = foo.GetBuffer();
   ```
   should be transpiled to
   ```php
   $foo = new Foo();
   $buffer = &$foo->getBuffer();
   ```
   Not sure how that will behave in `foreach` though, because `foreach ( (&$foo->get_buffer()) as $value)` is a syntax error. Perhaps in `foreach` we don't have to use `&` when dereferencing.
4. In PHP, it's standard to use `camelCase`, so the method probably should be `getBuffer(): array`, not `get_buffer`.
5. PHP also has modifiers, so I think we should explicitly add `public` modifier in PHP for `public` and `internal` Fusion methods. Of course, when the modifier is omitted, it's public by default in PHP, but I still think it's a good idea to explicitly specify, since all projects I saw in PHP had `public` specified explicitly.
6. I'm thinking - is the transpiled code to ever be edited by users? I'm asking - do you expect users to write code in Fusion, then transpile it to PHP, and the edit the PHP code itself? Because if so, then using references is not really a standard way. I understand that it's used to mimic the `List<>` behaviour of Java/C#, and that's probably the only way to do it, but PHP developer would not return that array as reference. If transpiled PHP code is only ever supposed to be executed, then it's no problem. But if it's supposed to be editable as a general-purpose code, then perhaps we can put more thought into it?

**Comment by pfusik** [comment]
Thanks, that was very helpful!

The emitted code is not meant to be edited, but it is meant to be human readable.

Array references aren't just a "Javism", they have different semantics, e.g.

```php
$foo = new Foo; // or new Foo() ?
$foo->push(5);
$buf = &$foo->getBuffer();
$buf[0] = 7;
$foo->push(10);
assert($buf == [7, 10]);
```

What's the naming convention for `const`s and enum values?

**Reply by danon** [reply]
> What's the naming convention for consts and enum values?

`ALL_CAPS`.

> Array references aren't just a "Javism", they have different semantics, e.g.
> 
> ```
> $foo = new Foo; // or new Foo() ?
> $foo->push(5);
> $buf = &$foo->getBuffer();
> $buf[0] = 7;
> $foo->push(10);
> assert($buf == [7, 10]);
> ```

So I reckon. That should be fine.

My expectation would be to see `new Foo()`.

> The emitted code is not meant to be edited, but it is meant to be human readable.

Okay, then references are a good way to go I think.

However, I see another problem with `array`. If the generic in `Foo` should be `List`, then that would be the Fusion code?

```
class Foo
{
	List<List<byte>>() Buffer;
	internal void Push!(List<byte> b)
	{
		Buffer.Add(b);
	}
	public List<List<byte>> GetBuffer() => Buffer;
}
```

Because, if so, then method accepting the argument should probably be also `&`:
```php
class Foo
{
    private array $buffer = [];
    function push(array &$b): void
    {
        $this->buffer[] = &$b;
    }
    function &getBuffer(): array
    {
        return $this->buffer;
    }
}
```

Because of:
```php
<?php

class Foo
{
    private array $buffer = [];
    function push(array &$b): void
    {
        $this->buffer[] = &$b;
    }
    function &getBuffer(): array
    {
        return $this->buffer;
    }
}

$fruits = ['apple', 'orange', 'pear'];
$colors = ['red', 'yellow', 'green'];

$collections = new Foo();
$collections->push($fruits);
$collections->push($colors);

foreach ($collections->getBuffer() as &$list) { # notice &
    $list[0] = 'overridden'; # this should override an item in nested list
}

var_dump($fruits);
var_dump($colors);
```

But that yields a problem, so that if you use `&$list` in `foreach` in PHP, then it's possible not only to override the contents of the array, but also the reference itself.
```
foreach ($collections->getBuffer() as &$list) {
  $list = ['new', 'list']; # simple variable assignment, would modify the contents of Foo
}
```

Not sure there is a way to sidestep it, other than use alias variables.

So that the code in fusion:
```
# fusion
List<string> fruits = new List<string>(); 
fruits.add("apple");
fruits.add("orange");

foo = new Foo();
foo.add(fruits);

foreach (buffer = foo.GetBuffer()) { # i'm not sure this is the correct fusion syntax
  buffer[0] = 'new item';
  buffer = new List<>();
  buffer.add('something else');
}
```
that would probably have to be transpiled to:
```
# fusion
$fruits = ["apple", "orange"];

$foo = new Foo();
$foo->add($fruits);

foreach (foo.GetBuffer() as &$buffer) { 
  $buffer[0] = 'new item';
  $buffer_alias = new List<>();
  $buffer_alias->add('something else');
}
```

The transpiler would have to differentiate using a reference `&$buffer; &buffer[0] = 2` and reassigning the reference `&$buffer; $buffer = 'something else';`. In the first case it should use it, in the other case it should probably use an alias variable. 

An alternative would be to use wrapper `List` class, but I'm not sure whether that's a right approach.`array` are de-facto standard for collections in PHP.

Same should go for simple variable reassignment

```
# fusion
List<byte> list = new List<>();
new_list = list;
```
should be transpiled to
```php
$list = ['one', 'two'];
$newList = &$list;
```

### Conclusion

I think we should use `&` in return types, arguments and referencing variables, because of passing a mutable object. In case of primitives (`int`, `string`, `byte` we don't need that). So: arguments, variables, return types, loops, any kind of variable access should be with `&` to mimic object semantics of the array.

But - usage of reference variables should only be "straightforward" when we *access* the variable, so `$var[0]`, `$var->field`. When we reassign it (`$var = `), then the transpiler should assign the lefthandside of the operant to a different variable.

Because, for example
```
# Java
list = object.list;
list[0] = 'new value'; // edits the list in object
list = new List(); // list in object remains intact
```
but in PHP
```
# PHP
$list = &object->list;
$list[0] = 'new value'; // edits the list in object
$list = ['new'];  // this line will actually override the list in object
```

So to fix it, we need to assign the new value to a temporary variable and use that aliased variable, or we can set the new reference.
```
# Java
list = object.list;
list[0] = 'new value'; // edits the list in object
list = new List() { "new value" } ; // list in object remains intact
```
could be transpiled to 
```
# PHP
$list = &object->list;
$list[0] = 'new value'; // edits the list in object
$temporary = ['new'];
$list = &temporary;  // now list in object remains intact
```

Same trick would have to be used for using return values and foreach variables, and all places we are dereferencing non-primitive type.

**Reply by danon** [reply]
So actually, reference variables in PHP have two effects:

1. They mimic object semantics similar to Java,  C# (so the collection is actually a reference)
2. They change variables into a kind of reactive programming style (so that reassigning local variables can change the contents of object properties).

We need to first behaviour, and deal somehow with the second.

Altought I'm pretty sure we only need this wierd behaviour for `array`. Primitives don't need it, and objects do it by default. That's why a possible alternative would be to use custom `List` object, but that would yield less php-ish code. `array` is more php-ish, but gives the troubles with references and reassignment.
